File: ios/Classes/LightweightBarcodeScannerPlugin.swift
import AVFoundation
import Flutter
import UIKit

/// Method-channel front end.
///
/// It owns the scanner sessions and nothing else: camera work lives in
/// `ScannerSession` and decoding lives in the shared C++ core.
public class LightweightBarcodeScannerPlugin: NSObject, FlutterPlugin {
  private let registry: FlutterTextureRegistry
  private let messenger: FlutterBinaryMessenger
  private var sessions: [Int: ScannerSession] = [:]
  private var nextSessionId = 1

  init(registry: FlutterTextureRegistry, messenger: FlutterBinaryMessenger) {
    self.registry = registry
    self.messenger = messenger
    super.init()
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "com.enver.lightweight_barcode_scanner/methods",
      binaryMessenger: registrar.messenger()
    )
    let instance = LightweightBarcodeScannerPlugin(
      registry: registrar.textures(),
      messenger: registrar.messenger()
    )
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.publish(instance)
  }

  public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
    releaseAllSessions()
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let arguments = call.arguments as? [String: Any] ?? [:]

    switch call.method {
    case "checkPermission":
      result(LightweightBarcodeScannerPlugin.permissionStatus())

    case "requestPermission":
      requestPermission(result: result)

    case "create":
      let id = nextSessionId
      nextSessionId += 1
      sessions[id] = ScannerSession(
        sessionId: id,
        registry: registry,
        messenger: messenger,
        configuration: ScannerConfiguration.from(arguments)
      )
      result(id)

    case "start":
      withSession(arguments, result) { session in
        session.start { outcome in
          switch outcome {
          case .success(let preview): result(preview)
          case .failure(let error): result(Self.flutterError(error))
          }
        }
      }

    case "pause":
      withSession(arguments, result) { $0.pause(); result(nil) }

    case "resume":
      withSession(arguments, result) { $0.resume(); result(nil) }

    case "stop", "dispose":
      if let id = arguments["sessionId"] as? Int {
        sessions.removeValue(forKey: id)?.release()
      }
      result(nil)

    case "setTorch":
      withSession(arguments, result) { session in
        self.attempt(result) {
          try session.setTorch(arguments["enabled"] as? Bool ?? false)
        }
      }

    case "setZoom":
      withSession(arguments, result) { session in
        self.attempt(result) {
          try session.setZoom((arguments["zoom"] as? NSNumber)?.doubleValue ?? 1)
        }
      }

    case "setFocusPoint":
      withSession(arguments, result) { session in
        self.attempt(result) {
          try session.setFocusPoint(
            x: (arguments["x"] as? NSNumber)?.doubleValue,
            y: (arguments["y"] as? NSNumber)?.doubleValue
          )
        }
      }

    case "switchCamera":
      withSession(arguments, result) { session in
        session.switchCamera(facing: arguments["facing"] as? String ?? "back") {
          outcome in
          switch outcome {
          case .success(let preview): result(preview)
          case .failure(let error): result(Self.flutterError(error))
          }
        }
      }

    case "setFormats":
      withSession(arguments, result) { session in
        session.setFormats((arguments["formats"] as? NSNumber)?.uint32Value ?? 0)
        result(nil)
      }

    case "setScanRegion":
      withSession(arguments, result) { session in
        var region: CGRect?
        if let values = arguments["scanRegion"] as? [NSNumber], values.count >= 4 {
          region = CGRect(
            x: values[0].doubleValue, y: values[1].doubleValue,
            width: values[2].doubleValue, height: values[3].doubleValue)
        }
        session.setScanRegion(region)
        result(nil)
      }

    case "setDuplicateFilter":
      withSession(arguments, result) { session in
        session.setDuplicateFilter(
          millis: (arguments["duplicateFilterMillis"] as? NSNumber)?.doubleValue ?? 0)
        result(nil)
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - helpers

  private func withSession(
    _ arguments: [String: Any],
    _ result: @escaping FlutterResult,
    _ body: (ScannerSession) -> Void
  ) {
    guard let id = arguments["sessionId"] as? Int, let session = sessions[id] else {
      result(
        FlutterError(
          code: "invalidState",
          message: "This scanner session is no longer available.",
          details: nil))
      return
    }
    body(session)
  }

  private func attempt(_ result: @escaping FlutterResult, _ body: () throws -> Void) {
    do {
      try body()
      result(nil)
    } catch {
      result(Self.flutterError(error))
    }
  }

  private static func flutterError(_ error: Error) -> FlutterError {
    if let scannerError = error as? ScannerError {
      return FlutterError(
        code: scannerError.code,
        message: scannerError.errorDescription,
        details: nil)
    }
    return FlutterError(
      code: "unknown", message: error.localizedDescription, details: nil)
  }

  private static func permissionStatus() -> String {
    switch AVCaptureDevice.authorizationStatus(for: .video) {
    case .authorized: return "granted"
    case .denied: return "permanentlyDenied"
    case .restricted: return "restricted"
    case .notDetermined: return "denied"
    @unknown default: return "denied"
    }
  }

  private func requestPermission(result: @escaping FlutterResult) {
    switch AVCaptureDevice.authorizationStatus(for: .video) {
    case .authorized:
      result("granted")
    case .notDetermined:
      AVCaptureDevice.requestAccess(for: .video) { granted in
        DispatchQueue.main.async {
          // A denial at the system prompt is final: iOS never shows it twice.
          result(granted ? "granted" : "permanentlyDenied")
        }
      }
    case .denied:
      result("permanentlyDenied")
    case .restricted:
      result("restricted")
    @unknown default:
      result("denied")
    }
  }

  private func releaseAllSessions() {
    sessions.values.forEach { $0.release() }
    sessions.removeAll()
  }
}

File: android/src/main/kotlin/com/enver/lightweight_barcode_scanner/LightweightBarcodeScannerPlugin.kt
package com.enver.lightweight_barcode_scanner

import android.app.Activity
import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry

/**
 * Method-channel front end.
 *
 * It owns the scanner sessions and nothing else: camera work lives in
 * [ScannerSession] and decoding lives in the shared C++ core.
 */
class LightweightBarcodeScannerPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var messenger: BinaryMessenger
    private lateinit var textureRegistry: TextureRegistry

    private val permissions = CameraPermissions()
    private val sessions = mutableMapOf<Int, ScannerSession>()
    private var nextSessionId = 1

    private var activityBinding: ActivityPluginBinding? = null
    private val activity: Activity? get() = activityBinding?.activity

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        messenger = binding.binaryMessenger
        textureRegistry = binding.textureRegistry
        channel = MethodChannel(messenger, METHOD_CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        releaseAllSessions()
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activityBinding = binding
        binding.addRequestPermissionsResultListener(permissions)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) =
        onAttachedToActivity(binding)

    override fun onDetachedFromActivityForConfigChanges() = onDetachedFromActivity()

    override fun onDetachedFromActivity() {
        activityBinding?.removeRequestPermissionsResultListener(permissions)
        activityBinding = null
        // The activity is going away and with it the surfaces the camera is
        // drawing into. Hold on to nothing.
        releaseAllSessions()
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        try {
            when (call.method) {
                "checkPermission" -> result.success(permissions.status(context))

                "requestPermission" -> permissions.request(activity, context) { status ->
                    result.success(
                        if (status == CameraPermissions.GRANTED) status
                        else permissions.statusAfterDenial(activity),
                    )
                }

                "create" -> {
                    val arguments = call.arguments as? Map<*, *>
                        ?: throw IllegalArgumentException("create requires arguments")
                    val id = nextSessionId++
                    sessions[id] = ScannerSession(
                        id = id,
                        context = context,
                        textureRegistry = textureRegistry,
                        messenger = messenger,
                        configuration = ScannerConfiguration.fromMap(arguments),
                    )
                    result.success(id)
                }

                "start" -> session(call).start { outcome ->
                    outcome.fold(
                        onSuccess = result::success,
                        onFailure = { error ->
                            result.error(
                                "cameraInitializationFailed",
                                error.message ?: "Could not open the camera.",
                                null,
                            )
                        },
                    )
                }

                "pause" -> {
                    session(call).pause()
                    result.success(null)
                }

                "resume" -> {
                    session(call).resume()
                    result.success(null)
                }

                "stop", "dispose" -> {
                    val id = sessionId(call)
                    sessions.remove(id)?.release()
                    result.success(null)
                }

                "setTorch" -> {
                    session(call).setTorch(call.argument<Boolean>("enabled") == true)
                    result.success(null)
                }

                "setZoom" -> {
                    val zoom = call.argument<Double>("zoom")
                        ?: throw IllegalArgumentException("setZoom requires a zoom")
                    session(call).setZoom(zoom.toFloat())
                    result.success(null)
                }

                "setFocusPoint" -> {
                    session(call).setFocusPoint(
                        call.argument<Double>("x")?.toFloat(),
                        call.argument<Double>("y")?.toFloat(),
                    )
                    result.success(null)
                }

                "switchCamera" -> {
                    val facing = call.argument<String>("facing") ?: "back"
                    session(call).switchCamera(facing) { outcome ->
                        outcome.fold(
                            onSuccess = result::success,
                            onFailure = { error ->
                                result.error(
                                    "cameraUnavailable",
                                    error.message ?: "Could not switch cameras.",
                                    null,
                                )
                            },
                        )
                    }
                }

                "setFormats" -> {
                    session(call).setFormats(call.argument<Int>("formats") ?: 0)
                    result.success(null)
                }

                "setScanRegion" -> {
                    session(call).setScanRegion(
                        ScannerConfiguration.scanAreaFrom(
                            call.argument<List<Double>>("scanRegion"),
                        ),
                    )
                    result.success(null)
                }

                "setDuplicateFilter" -> {
                    session(call).setDuplicateFilter(
                        call.argument<Int>("duplicateFilterMillis")?.toLong() ?: 0L,
                    )
                    result.success(null)
                }

                else -> result.notImplemented()
            }
        } catch (error: UnsupportedOperationException) {
            result.error("unsupportedOperation", error.message, null)
        } catch (error: IllegalStateException) {
            result.error("invalidState", error.message, null)
        } catch (error: Throwable) {
            result.error("unknown", error.message ?: error.toString(), null)
        }
    }

    private fun sessionId(call: MethodCall): Int =
        call.argument<Int>("sessionId")
            ?: throw IllegalArgumentException("${call.method} requires a sessionId")

    private fun session(call: MethodCall): ScannerSession =
        sessions[sessionId(call)]
            ?: throw IllegalStateException("This scanner session is no longer available.")

    private fun releaseAllSessions() {
        sessions.values.forEach(ScannerSession::release)
        sessions.clear()
    }

    private companion object {
        const val METHOD_CHANNEL = "com.enver.lightweight_barcode_scanner/methods"
    }
}
